drivers/sensor/lsm6dsl/lsm6dsl.c

```diff
@@ -295,6 +354,84 @@ static void lsm6dsl_gyro_channel_get_temp(struct sensor_value *val,
 }
 #endif
 
+#if defined(CONFIG_LSM6DSL_EXT0_LIS2MDL)
+static inline void lsm6dsl_magn_convert(struct sensor_value *val, int raw_val,
+					float sensitivity)
+{
+	double dval;
+
+	/* Sensitivity is exposed in mgauss/LSB */
+	dval = (double)(raw_val * sensitivity);
+	val->val1 = (s32_t)dval / 1000000;
+	val->val2 = (s32_t)dval % 1000000;
+}
+
+static inline int lsm6dsl_magn_get_channel(enum sensor_channel chan,
+					   struct sensor_value *val,
+					   struct lsm6dsl_data *data)
+{
+	switch (chan) {
+	case SENSOR_CHAN_MAGN_X:
+		lsm6dsl_magn_convert(val,
+				     data->magn_sample_x,
+				     data->magn_sensitivity);
+		break;
+	case SENSOR_CHAN_MAGN_Y:
+		lsm6dsl_magn_convert(val,
+				     data->magn_sample_y,
+				     data->magn_sensitivity);
+		break;
+	case SENSOR_CHAN_MAGN_Z:
+		lsm6dsl_magn_convert(val,
+				     data->magn_sample_z,
+				     data->magn_sensitivity);
+		break;
+	case SENSOR_CHAN_MAGN_XYZ:
+		lsm6dsl_magn_convert(val,
+				     data->magn_sample_x,
+				     data->magn_sensitivity);
+		lsm6dsl_magn_convert(val + 1,
+				     data->magn_sample_y,
+				     data->magn_sensitivity);
+		lsm6dsl_magn_convert(val + 2,
+				     data->magn_sample_z,
+				     data->magn_sensitivity);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int lsm6dsl_magn_channel_get(enum sensor_channel chan,
+				    struct sensor_value *val,
+				    struct lsm6dsl_data *data)
+{
+	return lsm6dsl_magn_get_channel(chan, val, data);
```

Comment:
Why do you wrap `lsm6dsl_magn_get_channel` with another function?

Reply:
Well, actually we can set dynamically the magn attribute as well. Thanks to mention it, because in fact I had forgotten about this change that I had in mind. What I would need to do is adding the MAGN case in attr_set routine. In sensorhub file there should already be all the pertinent internal routines to program the external sensor.

Reply:
I just wrote the code, but it is not ready yet for submission. It seems that I need to think more about it and when I require to turn on/off the sensorhub. For the moment I will just provide the dynamic capability for lis2mdl only, leaving the lps22hb (not on ArgonKey) for the future.